server/chat-plugins/teams.ts

```diff
@@ -0,0 +1,686 @@
+			}
+			return buf;
+		},
+		async filtered(query, user, connection) {
+			const type = query.shift() || "";
+			TeamsHandler.validateAccess(connection);
+			let count = Number(query.shift()) || 50;
+			if (count > MAX_TEAMS) count = MAX_TEAMS;
+			let teams: StoredTeam[] = [], title = '';
+			const buttons: {[k: string]: string} = {
+				views: `<button class="button" name="send" value="/teams mostviews">Sort by most views</button>`,
+				latest: `<button class="button" name="send" value="/teams latest">Sort by most recent upload</button>`,
+			};
+			switch (type) {
+			case 'views':
+				this.title = `[Most Viewed Teams]`;
+				teams = await TeamsHandler.query(
+					`SELECT * FROM teams WHERE private = false ORDER BY views DESC LIMIT $1`, [count]
+				);
+				title = `Most viewed teams:`;
+				delete buttons.views;
+				break;
+			default:
+				this.title = `[Latest Teams]`;
+				teams = await TeamsHandler.query(
+					`SELECT * FROM teams WHERE private != true ORDER BY date DESC LIMIT $1`, [count]
+				);
+				title = `Recently uploaded teams:`;
+				delete buttons.latest;
+				break;
+			}
+			let buf = `<div class="ladder pad"><h2>${title}</h2>${refresh(this)}`;
+			buf += Object.values(buttons).join('<br />');
+			buf += `<hr />`;
+			buf += teams.map(team => TeamsHandler.preview(team, user)).join('<hr />');
+			buf += `</div>`;
+			return buf;
+		},
+		async view(query, user, connection) {
+			TeamsHandler.validateAccess(connection);
+			const teamid = parseInt(query.shift() || "");
+			this.title = `[View Team]`;
+			if (!teamid) {
+				throw new Chat.ErrorMessage(`Invalid team ID.`);
+			}
+			const team = await TeamsHandler.get(teamid);
+			if (!team) {
+				this.title = `[Invalid Team]`;
+				return this.errorReply(`No team with the ID ${teamid} was found.`);
+			}
+			this.title = `[Team] ${team.teamid}`;
+			if (user.id !== team.ownerid) {
+				void TeamsHandler.updateViews(team.teamid);
+			}
+			return `<div class="ladder pad">` + TeamsHandler.renderTeam(team, user) + "</div>";
+		},
+		upload(query, user, connection) {
+			TeamsHandler.validateAccess(connection);
+			this.title = `[Upload Team]`;
+			let buf = `<div class="ladder pad"><h2>Upload a team</h2>${refresh(this)}<hr />`;
+			// let [teamName, formatid, rawPrivacy, rawTeam] = Utils.splitFirst(target, ',', 4);
+			buf += `<form data-submitsend="/teams save {name},{format},{privacy},{team}">`;
+
+			buf += `<strong>What's the name of the team?</strong><br />`;
+			buf += `<input name="name" /><br />`;
+
+			buf += `<strong>What's the team's format?</strong><br />`;
+			buf += `<input name="format" /><br />`;
+
+			buf += `<strong>Should the team be private? (yes/no)</strong><br />`;
+			buf += `<select name="privacy" /><option value="1">Yes</option><option value="0">No</option></select><br />`;
+
+			buf += `<strong>Provide the team:</strong><br />`;
```

Comment:
There should be a way to paste a folder worth of teams which is useful for [team dumps](https://www.smogon.com/forums/threads/smogon-premier-league-xiv-sv-ou-discussion.3714163/page-2#post-9575543).